chisel/src/driver.rs

```diff
@@ -0,0 +1,544 @@
+    snip::Snip,
+    trimexports::TrimExports,
+    trimstartfunc::TrimStartFunc,
+    verifyexports::VerifyExports,
+    verifyimports::VerifyImports,
+    Module,
+    ModulePreset,
+    ModuleTranslator,
+    ModuleValidator,
+};
+
+/// State machine implementing the main chisel execution loop. Consumes ChiselConfig and returns
+/// ChiselResult, with an intermediate state returned to allow error handling.
+pub struct ChiselDriver<'a> {
+    config: &'a mut ChiselConfig,
+    state: DriverState,
+}
+
+/// The state of the chisel driver.
+pub enum DriverState {
+    Ready,
+    Error(DriverError, ChiselResult),
+    Done(ChiselResult),
+}
+
+pub enum DriverError {
+    /// A module or ruleset is missing a required field. Left-hand is the config object name,
+    /// right-hand is the missing field.
+    MissingRequiredField(String, String),
+    /// The module name was not successfully resolved to an existing chisel module. Left-hand is
+    /// the config object name, right-hand is the invalid module name.
+    ModuleNotFound(String, String),
+    /// A configuration value is of incorrect type or invalid value. Left-hand is the config object
+    /// name, right-hand is the name of the invalid field.
+    InvalidField(String, String),
+    /// A canonicalized path was generated unsuccessfully. Left-hand is the config object name,
+    /// right-hand is the invalid path.
+    PathResolution(String, String),
+    /// An internal error occurred. Field 0 is the config object, during the execution of which the error occurred.
+    /// Field 1 is an additional informational message. Field 2 is the error generated.
+    Internal(String, String, Box<dyn Error>),
+}
+
+impl<'a> ChiselDriver<'a> {
+    pub fn new(config: &'a mut ChiselConfig) -> Self {
+        ChiselDriver {
+            config,
+            state: DriverState::Ready,
+        }
+    }
+
+    pub fn fire(&mut self) -> &DriverState {
+        match &mut self.state {
+            DriverState::Ready => {
+                self.fire_internal();
+            }
+            DriverState::Error(_, _) => {
+                // Fire anyway. The previous successful results are preserved, but the failed
+                // result and accompanying error are discarded.
+                self.fire_internal();
+            }
+            DriverState::Done(_) => {
+                panic!("fire() called on a completed driver");
+            }
+        }
+        &self.state
+    }
+
+    fn fire_internal(&mut self) {
+        let mut results = match &mut self.state {
+            DriverState::Ready => ChiselResult::new(),
+            DriverState::Error(_, r) => r.clone(),
+            DriverState::Done(_) => panic!("fire_internal() called on a completed driver"),
+        };
+
+        // Consume the rulesets in the configuration and execute each one.
+        while let Some((name, mut ruleset)) = self.config.rulesets_mut().pop_front() {
+            let mut rs_result = RSResult::new(name.clone());
```

Comment:
Yeah really want to get this "rs" stuff renamed to ruleset throughout, because it is easier to understand that way.